content/developer/execution_model.md

```diff
@@ -0,0 +1,137 @@
+1. The dispatcher process:
+    - in case of a regular, non-force, cancel: waits for the workflow function to finish
+    - in case of a force-cancel: does not wait for the workflow function to finish
+    - sets the execution state to CANCELLED
+1. Workflow message is acked, no response is written
+
+
+## Kill-cancelling an execution
+
+To kill-cancel an execution, the user sends a POST request with the parameter "action" set to "kill".
+
+1. The restservice updates the execution state to CANCELLED.
+1. The restservice sends a "kill-execution" message to the management queue, containing at least the following fields:
+
+        {
+          'service_task': {
+              'task_name': 'cancel-workflow',
+              'kwargs': {
+                  'execution_id': execution id,
+                  'tenant': {
+                      'name': tenant name
+                  },
+                  'execution_token': execution token
+              }
+          }
+        }
+
+1. The management worker receives the "kill-execution" message.
+1. The workflow subprocess is killed by sending SIGTERM, and 5 seconds later, SIGKILL.
+1. The management worker sends a "kill-operation" message to every agent in the tenant of the execution. The "kill-operation" message contains at least the following fields:
+
+        {
+          'service_task': {
+              'task_name': 'cancel-operation',
+              'kwargs': {
+                  'execution_id': execution id
+              }
+          }
+        }
+
+1. Every agent receives the message, and kills every running operation subprocess of the cancelled execution by sending SIGTERM, and 5 seconds later, SIGKILL
+
+
+## Resuming an execution
+
+To kill-cancel an execution, the user sends a POST request with the parameter "action" set to "resume" or "force-resume".
+Force-resume is labeled `cfy executions resume --reset-operations` in the CLI.
+
+Only STARTED, CANCELLED and FAILED executions can be resumed, and only CANCELLED and FAILED executions can be force-resumed.
+Resuming STARTED executions is allowed to help restore "stuck" executions; to force-resume a STARTED execution, it should be cancelled first (if it is truly "stuck", it will have to be kill-cancelled).
+
+1. The restservice updates the execution object with:
+    - a new execution token
+    - ended_at date set to null
+    - status set to STARTED
+1. Update stored operations: operations in state RESCHEDULED or FAILED are set to PENDING, and their current_retries count is reset to 0.
+1. In case of a force-resume, also do the previous step for operations in state STARTED, SENT, or SENDING.
+1. Start running the execution as in the "Running an execution" section, with the exception that:
+    - where the management worker sends "start-operation" messages, if the operation is already in a SENT or STARTED state, the message is not sent. Instead, the management worker proceeds to the next step, and waits for a response.
```

Comment:
I assume there's more to come, given the 'proceeds to the next step' line?

Reply:
no, note that this is under a `Start running the execution as in the "Running an execution" section, with the exception that`. So, proceeds to the next step, of that other section. Reworded a bit